hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -1597,6 +1597,137 @@ private class Emit(
+          case PInt64(_) => (a: Code[_], b: Code[_]) => (coerce[Long](a) + coerce[Long](b)).asInstanceOf[Code[Any]]
+          case PFloat32(_) => (a: Code[_], b: Code[_]) => (coerce[Float](a) + coerce[Float](b)).asInstanceOf[Code[Any]]
+          case PFloat64(_) => (a: Code[_], b: Code[_]) => (coerce[Double](a) + coerce[Double](b)).asInstanceOf[Code[Any]]
+        }
+
+        val elementZero = lPType.elementType match {
+          case PInt32(_) => const(0).asInstanceOf[Code[Any]]
+          case PInt64(_) => const(0L).asInstanceOf[Code[Any]]
+          case PFloat32(_) => const(0.0f).asInstanceOf[Code[Any]]
+          case PFloat64(_) => const(0.0).asInstanceOf[Code[Any]]
+        }
+
+        val eVti = typeToTypeInfo(lPType.elementType.virtualType)
+
+        val emitter = new NDArrayEmitter(mb, outputPType.nDims, unifiedShapeArray, lPType.shape.pType, lPType.elementType, setup) {
+          override def outputElement(idxVars: Array[Code[Long]]): Code[_] = {
+            val seqIdxVars = idxVars.toSeq
+            val element = coerce[Any](mb.newField("matmul_element")(eVti))
+            val kLocal = mb.newLocal[Long]
+            val k = kLocal.load()
+
+            val (lIdxVars: Array[Code[Long]], rIdxVars: Array[Code[Long]]) = (lNDims, rNDims) match {
+              case (1, 1) => (Array(k), Array(k))
+              case (1, _) =>
+                val stackDims :+ m = seqIdxVars
+
+                val rStackVars = NDArrayEmitter.zeroBroadcastedDims(stackDims.toArray, rightBroadcastFlags)
+                (Array(k), rStackVars :+ k :+ m)
+              case (_, 1) =>
+                val stackDims :+ n = seqIdxVars
+
+                val lStackVars = NDArrayEmitter.zeroBroadcastedDims(stackDims.toArray, leftBroadcastFlags)
+                (lStackVars :+ n :+ k, Array(k))
+              case (_, _) => {
+                val stackDims :+ n :+ m = seqIdxVars
+                val lStackVars = NDArrayEmitter.zeroBroadcastedDims(stackDims.toArray, leftBroadcastFlags)
+                val rStackVars = NDArrayEmitter.zeroBroadcastedDims(stackDims.toArray, rightBroadcastFlags)
+
+                (lStackVars :+ n :+ k, rStackVars :+ k :+  m)
+              }
+            }
+
+
+
+            val lElem = Region.loadIRIntermediate(lPType.elementType)(lPType.getElementAddress2(lIdxVars, leftND, region, mb, lDataLocation.load(), lDataLength.load()))
+            val rElem = Region.loadIRIntermediate(rPType.elementType)(rPType.getElementAddress2(rIdxVars, rightND, region, mb, rDataLocation.load(), rDataLength.load()))
+            val maxK = mb.newLocal[Long]
+
+            val innerMethod = mb.fb.newMethod(eVti)
+
+            val loopCode = Code(
+              kLocal := 0L,
+              maxK := leftShapeArray(lNDims - 1),
+              element := elementZero,
+              //time := Code.timeMillis(),
+              Code.whileLoop(k < maxK,
```

Comment:
don't `kLocal` and `maxK` belong to the wrong methodbuilder? they are created with `mb.newLocal` but this code block is emitted to `innerMethod`

Reply:
Yeah, this seems true. Surprised that it works

Reply:
i've had a lot of cases like this where it ends up overwriting old locals but still happens to work. makes me nervous

Reply:
Don't worry, will fix. Honestly I mostly just use fields everywhere and quietly await the day Code2 will make the world less confusing.